File: interpreted/python/fastapi/nanocode.py
#!/usr/bin/env python3
"""nanocode - minimal claude code alternative (FastAPI web UI)"""
# pip install fastapi uvicorn && uvicorn nanocode:app
from fastapi import FastAPI
from fastapi.responses import HTMLResponse
from pydantic import BaseModel
import glob, json, os, re, subprocess, urllib.request

KEY = os.environ.get("ANTHROPIC_API_KEY")
MODEL = os.environ.get("MODEL", "claude-opus-4-5")
messages = []


def _read(a):
    lines = open(a["path"]).read().splitlines()
    off = a.get("offset", 0)
    lim = a.get("limit", len(lines))
    return "\n".join(f"{off+i+1}| {l}" for i, l in enumerate(lines[off : off + lim]))


def _write(a):
    open(a["path"], "w").write(a["content"])
    return "ok"


def _edit(a):
    text = open(a["path"]).read()
    old, new = a["old"], a["new"]
    count = text.count(old)
    if not count:
        return "error: old_string not found"
    if not a.get("all") and count > 1:
        return f"error: old_string appears {count} times, use all=true"
    updated = text.replace(old, new) if a.get("all") else text.replace(old, new, 1)
    open(a["path"], "w").write(updated)
    return "ok"


def _glob(a):
    pat = (a.get("path", ".") + "/" + a["pat"]).replace("//", "/")
    files = sorted(glob.glob(pat, recursive=True))
    return "\n".join(files) or "none"


def _grep(a):
    regex = re.compile(a["pat"])
    hits = []
    for path in glob.glob(a.get("path", ".") + "/**", recursive=True):
        if os.path.isdir(path):
            continue
        try:
            for i, line in enumerate(open(path), 1):
                if regex.search(line):
                    hits.append(f"{path}:{i}:{line.rstrip()}")
        except Exception:
            continue
    return "\n".join(hits[:50]) or "none"


def _bash(a):
    run = subprocess.run(a["cmd"], shell=True, capture_output=True, text=True)
    return (run.stdout + run.stderr).strip() or "(empty)"


tools = {"read": _read, "write": _write, "edit": _edit, "glob": _glob, "grep": _grep, "bash": _bash}

schema = [
    {"name": "read", "description": "Read file", "input_schema": {"type": "object", "properties": {"path": {"type": "string"}, "offset": {"type": "integer"}, "limit": {"type": "integer"}}, "required": ["path"]}},
    {"name": "write", "description": "Write file", "input_schema": {"type": "object", "properties": {"path": {"type": "string"}, "content": {"type": "string"}}, "required": ["path", "content"]}},
    {"name": "edit", "description": "Replace text", "input_schema": {"type": "object", "properties": {"path": {"type": "string"}, "old": {"type": "string"}, "new": {"type": "string"}, "all": {"type": "boolean"}}, "required": ["path", "old", "new"]}},
    {"name": "glob", "description": "Find files", "input_schema": {"type": "object", "properties": {"pat": {"type": "string"}, "path": {"type": "string"}}, "required": ["pat"]}},
    {"name": "grep", "description": "Search files", "input_schema": {"type": "object", "properties": {"pat": {"type": "string"}, "path": {"type": "string"}}, "required": ["pat"]}},
    {"name": "bash", "description": "Run command", "input_schema": {"type": "object", "properties": {"cmd": {"type": "string"}}, "required": ["cmd"]}},
]

def ask():
    req = urllib.request.Request("https://api.anthropic.com/v1/messages",
        data=json.dumps({"model": MODEL, "max_tokens": 4096, "system": "Concise assistant", "messages": messages, "tools": schema}).encode(),
        headers={"Content-Type": "application/json", "anthropic-version": "2023-06-01", "x-api-key": KEY})
    return json.loads(urllib.request.urlopen(req).read())

app = FastAPI()

class Msg(BaseModel):
    msg: str

@app.get("/", response_class=HTMLResponse)
def index():
    return """<!DOCTYPE html><html><head><title>nanocode</title>
<style>body{font:14px monospace;background:#0d1117;color:#c9d1d9;max-width:800px;margin:0 auto;padding:20px}
#out{white-space:pre-wrap}form{display:flex}input{flex:1;background:#21262d;border:1px solid #30363d;color:#c9d1d9;padding:8px}
button{background:#238636;border:none;color:#fff;padding:8px 16px}</style></head>
<body><h2>🧬 nanocode (FastAPI)</h2><div id="out"></div>
<form onsubmit="send(event)"><input id="i" autofocus><button>→</button></form>
<script>async function send(e){e.preventDefault();const m=i.value;i.value='';out.innerHTML+='❯ '+m+'\\n';
const r=await(await fetch('/chat',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({msg:m})})).json();
out.innerHTML+=r.map(b=>b.type==='text'?'⏺ '+b.text:'⚙ '+b.name).join('\\n')+'\\n\\n';}</script></body></html>"""

@app.post("/chat")
def chat(body: Msg):
    global messages
    if body.msg == "/c": messages = []; return [{"type": "text", "text": "Cleared"}]
    messages.append({"role": "user", "content": body.msg})
    out = []
    while True:
        content = ask().get("content", [])
        for block in content:
            if block["type"] == "text": out.append({"type": "text", "text": block["text"]})
            if block["type"] == "tool_use":
                out.append({"type": "tool", "name": block["name"]})
                result = tools[block["name"]](block["input"])
                messages.extend([{"role": "assistant", "content": content}, {"role": "user", "content": [{"type": "tool_result", "tool_use_id": block["id"], "content": result}]}])
        if not any(b["type"] == "tool_use" for b in content): messages.append({"role": "assistant", "content": content}); break
    return out

File: interpreted/typescript/deno/nanocode.ts
#!/usr/bin/env -S deno run -A
// nanocode - minimal claude code alternative (Deno)
const KEY = Deno.env.get("OPENROUTER_API_KEY") || Deno.env.get("ANTHROPIC_API_KEY");
const API = Deno.env.get("API_URL") || (Deno.env.get("OPENROUTER_API_KEY") ? "https://openrouter.ai/api/v1/messages" : "https://api.anthropic.com/v1/messages");
const MODEL = Deno.env.get("MODEL") || (Deno.env.get("OPENROUTER_API_KEY") ? "anthropic/claude-opus-4" : "claude-sonnet-4-20250514");
const [R, B, D, C, G] = ["\x1b[0m", "\x1b[1m", "\x1b[2m", "\x1b[36m", "\x1b[32m"];

const tools: Record<string, [(a: any) => string|Promise<string>, string, object]> = {
  read: [(a) => Deno.readTextFileSync(a.path).split("\n").slice(a.offset||0,(a.offset||0)+(a.limit||Infinity)).map((l,i)=>`${(a.offset||0)+i+1}| ${l}`).join("\n"), "Read file", {path:"string",offset:"number?",limit:"number?"}],
  write: [(a) => (Deno.writeTextFileSync(a.path, a.content), "ok"), "Write file", {path:"string",content:"string"}],
  edit: [(a) => { const t=Deno.readTextFileSync(a.path); if(!t.includes(a.old)) return "error: not found"; if(!a.all&&t.split(a.old).length>2) return "error: not unique"; Deno.writeTextFileSync(a.path, a.all?t.replaceAll(a.old,a.new):t.replace(a.old,a.new)); return "ok"; }, "Edit file", {path:"string",old:"string",new:"string",all:"boolean?"}],
  glob: [(a) => [...Deno.readDirSync(a.path||".")].flatMap(e => e.isDirectory ? [] : [e.name]).slice(0,50).join("\n")||"none", "Find files", {pat:"string",path:"string?"}],
  grep: [(a) => { const re = new RegExp(a.pat); const walk = (d:string):string[] => { try { return [...Deno.readDirSync(d)].flatMap(e => e.isDirectory ? walk(`${d}/${e.name}`) : [`${d}/${e.name}`]); } catch { return []; }}; return walk(a.path||".").flatMap(f => { try { return Deno.readTextFileSync(f).split("\n").map((l,i)=>[f,i+1,l] as const).filter(([,,l])=>re.test(l)).map(([f,n,l])=>`${f}:${n}:${l}`); } catch { return []; }}).slice(0,50).join("\n")||"none"; }, "Search", {pat:"string",path:"string?"}],
  bash: [async (a) => new TextDecoder().decode((await new Deno.Command("sh", {args:["-c",a.cmd]}).output()).stdout), "Run command", {cmd:"string"}],
};

const schema = Object.entries(tools).map(([name, [_, desc, params]]) => ({
  name, description: desc,
  input_schema: { type: "object", properties: Object.fromEntries(Object.entries(params).map(([k,v]) => [k, {type:(v as string).replace("?","")==="number"?"integer":(v as string).replace("?","")}])), required: Object.entries(params).filter(([_,v]) => !(v as string).endsWith("?")).map(([k]) => k) }
}));

async function ask(messages: any[]) {
  const res = await fetch(API, {
    method: "POST", headers: { "Content-Type": "application/json", "anthropic-version": "2023-06-01", ...(Deno.env.get("OPENROUTER_API_KEY") ? { Authorization: `Bearer ${KEY}` } : { "x-api-key": KEY! }) },
    body: JSON.stringify({ model: MODEL, max_tokens: 8192, system: `Concise coding assistant. cwd: ${Deno.cwd()}`, messages, tools: schema })
  });
  return res.json();
}

console.log(`${B}nanocode${R} | ${D}${MODEL}${R}\n`);
const messages: any[] = [];
const buf = new Uint8Array(1024);

while (true) {
  await Deno.stdout.write(new TextEncoder().encode(`${B}\x1b[34m❯${R} `));
  const n = await Deno.stdin.read(buf);
  if (n === null) break;
  const input = new TextDecoder().decode(buf.subarray(0, n)).trim();
  if (!input) continue;
  if (input === "/q") break;
  if (input === "/c") { messages.length = 0; console.log(`${G}⏺ Cleared${R}`); continue; }
  
  messages.push({ role: "user", content: input });
  while (true) {
    const { content } = await ask(messages);
    const results: any[] = [];
    for (const block of content) {
      if (block.type === "text") console.log(`\n${C}⏺${R} ${block.text}`);
      if (block.type === "tool_use") {
        console.log(`\n${G}⏺ ${block.name}${R}(${D}${JSON.stringify(block.input).slice(0,50)}${R})`);
        const result = await tools[block.name][0](block.input);
        console.log(`  ${D}⎿ ${result.split("\n")[0].slice(0,60)}${R}`);
        results.push({ type: "tool_result", tool_use_id: block.id, content: result });
      }
    }
    messages.push({ role: "assistant", content });
    if (!results.length) break;
    messages.push({ role: "user", content: results });
  }
  console.log();
}
